File: sdk/java/client/src/main/java/io/apibrew/client/impl/ChannelEventPoller.java
package io.apibrew.client.impl;

import io.apibrew.client.ApiException;
import io.apibrew.client.Client;
import io.apibrew.client.model.Extension;
import kong.unirest.HttpResponse;
import kong.unirest.Unirest;
import lombok.Builder;
import lombok.extern.log4j.Log4j2;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;

import static io.apibrew.client.helper.EventHelper.shortInfo;

@Builder
@Log4j2
public class ChannelEventPoller implements AutoCloseable {

    private final Client client;

    private final String channelKey;
    private final String threadName;

    private final Consumer<Extension.Event> consumer;
    private final AtomicBoolean stillRunning = new AtomicBoolean(false);
    private final AtomicBoolean isStopped = new AtomicBoolean(false);

    private final Thread thread = new Thread(this::run);

    // runs in background
    public void start() {
        ensureCanRun();
        thread.setName(threadName == null ? "ChannelEventPoller" : threadName);
        thread.start();
    }

    private void ensureCanRun() {
        if (isStopped.get()) {
            throw new IllegalStateException("Poller is stopped");
        }

        if (stillRunning.get()) {
            throw new IllegalStateException("Poller is already running");
        }
    }

    // runs in foreground
    public void run() {
        ensureCanRun();
        stillRunning.set(true);

        while (isRunning()) {
            try {
                log.debug("Begin Polling channel: {}", channelKey);

                HttpResponse<?> result = Unirest.get(ClientImpl.Urls.eventsUrl(client.getUrl()) + "?channelKey=" + channelKey)
                        .connectTimeout(10 * 1000)
                        .headers(client.headers())
                        .asObject(response -> {
                            if (!isRunning()) {
                                return null;
                            }

                            if (response.getStatus() != 200) {
                                throw new ApiException(response.getStatusText());
                            }

                            BufferedReader br = new BufferedReader(new InputStreamReader(response.getContent()));

                            try {
                                while (isRunning()) {
                                    String line = br.readLine();

                                    if (line == null) {
                                        break;
                                    }

                                    Extension.Event event = ClientImpl.objectMapper.readValue(line, Extension.Event.class);

                                    if (event.getId().equals("heartbeat-message")) {
                                        log.debug("Received heartbeat message");
                                        continue;
                                    }

                                    log.debug("Received event: {}", shortInfo(event));

                                    consumer.accept(event);

                                    return line;
                                }
                            } catch (IOException e) {
                                throw new RuntimeException(e);
                            } finally {
                                try {
                                    br.close();
                                } catch (IOException e) {
                                    throw new RuntimeException(e);
                                }
                            }

                            return null;
                        });

                ClientImpl.ensureResponseSuccess(result);

                log.debug("Polling channel: {} complete", channelKey);
            } catch (Exception e) {
                log.error("Error polling channel: {}", channelKey, e);
                try {
                    Thread.sleep(1000);
                } catch (InterruptedException ex) {
                    throw new RuntimeException(ex);
                }
            }
        }
    }

    private boolean isRunning() {
        return !isStopped.get() && stillRunning.get();
    }

    @Override
    public void close() {
        isStopped.set(true);
        stillRunning.set(false);

        thread.interrupt();
    }
}

File: sdk/nodejs/src/impl/channel-event-poller.ts
import {Client} from '../client';
import {BooleanExpression, Code, Event} from "../model/extension";
import axios from "axios";
import {Urls} from "./client-impl";
import {ApiException} from "../api-exception";


export class ChannelEventPoller {

    constructor(
        private client: Client,
        private channelKey: string,
        private threadName: string,
        private consumer: (event: Event) => void,
    ) {
    }

    private stillRunning: boolean = false;
    private isStopped: boolean = false;

    public start() {
        this.ensureCanRun();
        setTimeout(() => this.run(), 0);
    }

    private ensureCanRun() {
        if (this.isStopped) {
            throw new Error("Poller is stopped");
        }

        if (this.stillRunning) {
            throw new Error("Poller is already running");
        }
    }

    private async run() {
        this.stillRunning = true;

        while (this.isRunning()) {
            try {
                const url = Urls.eventsUrl(this.client.getUrl()) + "?channelKey=" + this.channelKey

                console.log("Begin Polling channel: " + this.channelKey + " at: " + url);

                const result = await axios.get(url, {
                    headers: this.client.headers(),
                    validateStatus: (status) => true,
                    responseType: "stream",
                    timeout: 10 * 1000,
                });

                if (!this.isRunning()) {
                    return;
                }

                if (result.status !== 200) {
                    console.log("Polling channel: " + this.channelKey + " failed: " + result.status + ": " + result.statusText);
                    await new Promise(resolve => setTimeout(resolve, 1000));
                    continue;
                }

                const stream = result.data;

                // block until stream is closed

                let wc = 0;

                wc++;

                stream.on('data', (data: Buffer) => {
                    const event = JSON.parse(data.toString()) as Event;

                    if (event.id === "heartbeat-message") {
                        console.log("Received heartbeat message");
                    } else {
                        console.log("Received event: " + event);
                        this.consumer(event);
                    }
                });

                stream.on('end', () => {
                    console.log("stream done");
                    wc--;
                });

                stream.on('error', (err: any) => {
                    console.log("stream error: " + err);
                    wc--;
                });

                while (wc > 0) {
                    await new Promise(resolve => setTimeout(resolve, 100));
                }

                console.log("Polling channel: " + this.channelKey + " complete");
            } catch (e) {
                console.log("Error polling channel: " + this.channelKey + ": " + e);
                try {
                    await new Promise(resolve => setTimeout(resolve, 1000));
                } catch (ex: any) {
                    throw new Error(ex.message);
                }
            }
        }
    }

    private isRunning() {
        return !this.isStopped && this.stillRunning;
    }

    public close() {
        this.isStopped = true;
        this.stillRunning = false;
    }

}
